hpedockerplugin/cmd/cmd_deleteshare.py

```diff
@@ -21,20 +20,61 @@ def __init__(self, file_mgr, share_info):
 
     def execute(self):
         LOG.info("Delting share %s..." % self._share_info['name'])
+        # Most likely nothing got created at the backend when share is
+        # not in AVAILABLE state
+        if self._share_info['status'] != 'AVAILABLE':
+            self._delete_share_from_etcd(self._share_info['name'])
+            return json.dumps({u"Err": ''})
+
         with self._fp_etcd.get_fpg_lock(
                 self._backend, self._cpg_name, self._fpg_name):
             self._remove_quota()
             self._delete_share()
+
             # Decrement count only if it is Docker managed FPG
             if self._share_info.get('docker_managed'):
-                remaining_cnt = self._decrement_share_cnt()
-                if remaining_cnt == 0:
-                    self._delete_fpg()
+                self._decrement_share_cnt()
+
+            # If shares are not present on FPG after this delete, then
+            # delete the FPG too.
+            # WARNING: THIS WILL DELETE LEGACY FPG TOO IF IT BECOMES EMPTY
+            if not self._mediator.shares_present_on_fpg(self._fpg_name):
+                self._delete_fpg()
```

Comment:
Here we can check whether fpg is legacy or Docker created FPG by querying etcd data base. If FPG is legacy FPG we will delete only etcd entry if present, if it is docker fpg, only then we will delete etcd entry and real FPG from 3PAR. As discussed over the call we will take it as separate PR. Rest of changes looks fine for me. Hence will approve this pull request and will merge.

1. If share is on legacy FPG, FPG will not be deleted , this has to be done manually by Administrator.
2. IF docker created FPG has stores on it while deleting docker created last share FPG will not be deleted. It needs to be cleaned by Administrator.